cdap-watchdog/src/main/java/co/cask/cdap/logging/framework/LogFileManager.java

```diff
@@ -0,0 +1,207 @@
+   */
+  public LogFileOutputStream getLogFileOutputStream(LogPathIdentifier logPathIdentifier,
+                                                    long eventTimestamp) throws IOException {
+    LogFileOutputStream logFileOutputStream = outputStreamMap.get(logPathIdentifier);
+    if (logFileOutputStream == null) {
+      logFileOutputStream = createOutputStream(logPathIdentifier, eventTimestamp);
+    } else {
+      // rotate the file if needed (time has passed)
+      logFileOutputStream = rotateOutputStream(logFileOutputStream, logPathIdentifier, eventTimestamp);
+    }
+    return logFileOutputStream;
+  }
+
+  private LogFileOutputStream createOutputStream(final LogPathIdentifier identifier,
+                                                 long timestamp) throws IOException {
+    TimeStampLocation location = createLocation(identifier);
+    try {
+      fileMetaDataManager.writeMetaData(identifier, timestamp, location.getTimeStamp(), location.getLocation());
+    } catch (Throwable e) {
+      // delete created file as there was exception while writing meta data
+      Locations.deleteQuietly(location.getLocation());
+      throw new IOException(e);
+    }
+
+    LOG.info("Created Avro file at {}", location);
+    LogFileOutputStream logFileOutputStream = new LogFileOutputStream(
+      location.getLocation(), schema, syncIntervalBytes, new Closeable() {
+      @Override
+      public void close() throws IOException {
+        outputStreamMap.remove(identifier);
+      }
+    });
+
+    outputStreamMap.put(identifier, logFileOutputStream);
+    return logFileOutputStream;
+  }
+
+
+  private TimeStampLocation createLocation(LogPathIdentifier logPathIdentifier) throws IOException {
+    // if createNew fails, we retry after sleeping for a milli second as we use current timestamp for fileName.
+    // this retry should succeed on any potential conflicts, though the likelihood of conflict is very small.
+    TimeStampLocation location = getLocation(logPathIdentifier);
+
+    LOG.trace("Trying to create new file at Location {}", location);
+    // NOTE : we are creating a file before writing meta data, so if there are simultaneous write to same folder from
+    // multiple instance of log appender, they wouldn't overwrite the meta data and would fail early,
+    // so we can retry with different timestamp for file creation.
+    // however there is a possibility for the log.saver could crash after file was created
+    // but before meta data was written, log clean up should handle this scenario.
+    // log cleanup shouldn't rely on metadata table for cleaning up old files.
+    if (!location.getLocation().createNew()) {
+      LOG.trace("create new file at Location {} failed. Will retry after 1ms.", location);
+      try {
+        TimeUnit.MILLISECONDS.sleep(1);
+      } catch (InterruptedException ie) {
+        LOG.warn("Interrupted during sleep before create location");
+      }
+      createLocation(logPathIdentifier);
```

Comment:
Better not to use recursion. It can simply be:

```java
while (!location.getLocation().createNew()) {
  Uninterruptibles.sleepUninterruptibly(1L, TimeUnit.MILLISECONDS);
  location = getLocation(logPathIdentifier);
}
return location;
```